contracts/modules/VersionManager.sol

```diff
@@ -0,0 +1,296 @@
+// Copyright (C) 2018  Argent Labs Ltd. <https://argent.xyz>
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+import "./common/Utils.sol";
+import "../infrastructure/base/Owned.sol";
+import "../infrastructure/storage/ITransferStorage.sol";
+import "../infrastructure/storage/IGuardianStorage.sol";
+import "./common/IModule.sol";
+import "./common/BaseFeature.sol";
+
+/**
+ * @title VersionManager
+ * @notice Intermediate contract between features and wallets. VersionManager checks that a calling feature is
+ * authorised for the wallet and if so, forwards the call to it.
+ * @author Olivier VDB <olivier@argent.xyz>
+ */
+contract VersionManager is IVersionManager, IModule, BaseFeature, Owned {
+
+    bytes32 constant NAME = "VersionManager";
+
+    bytes4 constant internal ADD_MODULE_PREFIX = bytes4(keccak256("addModule(address,address)"));
+    bytes4 constant internal UPGRADE_WALLET_PREFIX = bytes4(keccak256("upgradeWallet(address)"));
+
+    // Last bundle version
+    uint256 public lastVersion;
+    // Current bundle version for a wallet
+    mapping(address => uint256) public walletVersions; // [wallet] => [version]
+    // Features per version
+    mapping(address => mapping(uint256 => bool)) public isFeatureInVersion; // [feature][version] => bool
+    // Features requiring initialization for a wallet
+    mapping(uint256 => address[]) public featuresToInit; // [version] => [features]
+
+    // Supported static call signatures
+    mapping(uint256 => bytes4[]) public staticCallSignatures; // [version] => [sigs]
+    // Features executing static calls
+    mapping(uint256 => mapping(bytes4 => address)) public staticCallExecutors; // [version][sig] => [feature]
```

Comment:
Given the current storage structure it is impossible to get the list of features from a version. If I remember correctly this was requested by the clients. We should check with @th3m477 if it is still the case.

Reply:
I added the list of features in a version to the `VersionAdded` event: `event VersionAdded(uint256 _version, address[] _features)`. Will check with clients if that's acceptable.

Reply:
Each app will have a version number associated with it, and as we can get version number from wallet we will assume everything else matches (no need to check specific modules).

How do I get the version number, given a wallet address? I can look at the public getter on `walletVersions`; does that mean VersionManager will not change?